HW2 part 2 - support subdomains

anton-ptashnik opened the pull request:
- [x] add support for subdomains
- [x] add tests to verify subdomain support 
- [x] add a migration script to verify ContractV2 can be deployed without storage corruption

Please refer to README for test run examples https://github.com/anton-ptashnik/web3-domainregistar/blob/hw2p2-support-subdomains/README.md

anton-ptashnik committed:
impl subdomain registration

anton-ptashnik committed:
impl price update for subdomains

anton-ptashnik committed:
impl per-owner balance accounting

anton-ptashnik committed:
refactor: dedicate a library for domain related code

anton-ptashnik committed:
make API external

anton-ptashnik committed:
code cleanup

anton-ptashnik committed:
add basic contract upgrade tests

anton-ptashnik committed:
add contract upgrade script

anton-ptashnik committed:
add a script for running contract upgrade tests

anton-ptashnik committed:
add a test to verify upgrade do not corrupt v1 functionality

anton-ptashnik committed:
fix a test script

anton-ptashnik committed:
fix test docs

anton-ptashnik committed:
add post-upgrade test to verify subdomains support

anton-ptashnik committed:
add a convenience command for running tests

anton-ptashnik committed:
move an upgrade testing script

anton-ptashnik committed:
add a post-upgrade coin withdrawal test

anton-ptashnik committed:
update datasets

anton-ptashnik committed:
fix issue preventing a contract owner to withdraw coins earned before upgrade

anton-ptashnik committed:
move auxiliary code to a separate file

anton-ptashnik committed:
rearrange code, remove extra vars

anton-ptashnik committed:
add docs for DomainUtils

anton-ptashnik committed:
refactor a script for running upgrade tests

anton-ptashnik committed:
refer to v1 v2 codebase via tags

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -4,35 +4,25 @@ pragma solidity ^0.8.24;
-
+import "./DomainUtils.sol";
 
 error AccessDenied(string description);
```
line 9 - 11
errors uot of contracts scope
// - 1 point

anton-ptashnik replied:
please clarify here. Do you mean a lib function should not revert byself and only communicate to the contract if call passed/failed?

liscody replied:
no 
it should have a place inside in contract scope 

contract DomainRegistar is Initializable {
    error ... 
}

anton-ptashnik replied:
moved error defs to a contract file, please check

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -4,35 +4,25 @@ pragma solidity ^0.8.24;
-        address payable owner;
-
-        /// @notice Price payed for domain registration
-        uint weiDomainPrice;
```
line 24  and other line with same logic 
    uint -> as a good practice, it is better to use uint256
    // - 1 point

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -4,35 +4,25 @@ pragma solidity ^0.8.24;
+        mapping(address => uint) shares;
     }
     
     // keccak256(abi.encode(uint256(keccak256("DomainRegistar.main")) - 1)) & ~bytes32(uint256(0xff));
```
line 29 
constant should be declared before contract variables
// - 1 point

IhorSolDevRD replied:
line 31 
function location 
Private functions should be declared after public functions
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -59,31 +49,48 @@ contract DomainRegistar is Initializable {
      */
     event PriceChanged(uint newPrice, uint oldPrice);
 
+    using DomainUtils for DomainUtils.DomainEntry;
```
line 52 
Library`s should be declared before contract variables
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -59,31 +49,48 @@ contract DomainRegistar is Initializable {
+    /**
+     * Return an address of a contract owner
+     */
+    function owner() external view returns (address) {
```
line 68 
it is easier to use oppenzeppelin`s Ownable contract upgradeable
// - 1 point

anton-ptashnik replied:
in my case usage of Ownable would introduce one extra state variable. Instead, I'm keeping contract owner address in root DomainEntry because in fact a contract creator is also an owner of an implicit root domain entry other top-level domains are branched from

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -93,60 +100,58 @@ contract DomainRegistar is Initializable {
      * Send all coins to the owner. Allowed for the owner only
      */
-    function withdraw() public {
+    function withdraw() external {
```
line 143
it is not necessary to load full storage to get shares
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistar.sol:
```diff
@@ -93,60 +100,58 @@ contract DomainRegistar is Initializable {
-        $.rootEntry.owner.transfer(address(this).balance);
+        uint balance = $.shares[msg.sender];
+        $.shares[msg.sender] = 0;
+        payable(msg.sender).transfer(balance);
```
line 147
call is better than transfer it use validation and reverts if the call fails
// - 1 point

anton-ptashnik replied:
used `call` revering with `Error` on failure. Used error type `Error` instead of the custom one for the sake of compatibility with v1 where `transfer` throws `Error` too

anton-ptashnik committed:
code reorganization according PR comments

- uint->uint256
- public API decl first, private - last
- remove extra var

anton-ptashnik committed:
resolved dependency conflict for hardhat-gas-reporter

anton-ptashnik committed:
Merge branch 'hw2p1-make-contract-upgradable' into hw2p2-support-subdomains

anton-ptashnik committed:
upd README

anton-ptashnik committed:
highlight steps for upgrade test

anton-ptashnik committed:
use call for coins transfer

anton-ptashnik committed:
code cleanup

- move error defs to a contract file
- do input validations in place, remove extra functions
